GVFS/GVFS.Common/Tracing/TelemetryDaemonEventListener.cs

```diff
@@ -96,39 +102,42 @@ private string SerializeMessage(TraceEventMessage message)
 
         private void SendMessage(string message)
         {
-            // Create pipe if this is the first message, or if the last connection broke for any reason
-            if (this.pipeClient == null)
+            lock (this.pipeLock)

```

Comment:
What is the performance impact of this lock?  Especially in things like updating placeholders, enumerating directories, parsing the index, etc.

Reply:
Unknown. I've not looked at it yet (this is still a draft PR for a reason).

It's also important to bear in mind that this event listener is only logging _telemetry_ information. I seriously hope we're not writing a trace message that's tagged at the 'telemetry' level for each and every file action.

Reply:
Plus even without locking, the fact we'd be writing a message over a pipe/socket for each and every file operation would most likely result in performance problems itself. We shouldn't be logging micro-events to telemetry.. only summary information.

Reply:
I agree.  This was more of a note to make sure we haven't introduced any regressions in performance with this change because with locking there is always that possibility.

Reply:
Is addressing performance issues in scope for this PR? We could cache the message and unblock the caller immediately. The cached messages can then be sent asynchronously.

Reply:
> We could cache the message and unblock the caller immediately. The cached messages can then be sent asynchronously.

At the minute we just drop failed messages on the floor upon failure.

Reply:
I've updated this to create a background thread and lock-free queue for posting the messages without holding up the calling threads.